File: backend/app/schema/analytics.py
"""
Pydantic schemas for analytics and statistics data.
"""
from pydantic import BaseModel
from typing import List, Optional, Dict, Any
from datetime import datetime


class TimeSeriesDataPoint(BaseModel):
    """Single data point in a time series."""
    date: str
    count: int
    value: Optional[float] = None


class UserStats(BaseModel):
    """User statistics."""
    total_users: int
    new_users_today: int
    new_users_this_week: int
    new_users_this_month: int
    active_users_today: int
    active_users_this_week: int
    active_users_this_month: int


class ActivityStats(BaseModel):
    """Activity statistics."""
    total_activities: int
    activities_today: int
    activities_this_week: int
    activities_this_month: int
    activities_by_action: Dict[str, int]
    activities_by_resource_type: Dict[str, int]


class OrganizationStats(BaseModel):
    """Organization statistics."""
    total_organizations: int
    total_memberships: int
    average_members_per_org: float
    organizations_created_today: int
    organizations_created_this_week: int
    organizations_created_this_month: int


class ActivityTimeSeries(BaseModel):
    """Time series data for activities."""
    period: str  # "day", "week", "month"
    data: List[TimeSeriesDataPoint]


class DashboardStats(BaseModel):
    """Combined dashboard statistics."""
    user_stats: UserStats
    activity_stats: ActivityStats
    organization_stats: OrganizationStats
    activity_timeline: ActivityTimeSeries


class UserGrowthTimeSeries(BaseModel):
    """Time series data for user growth."""
    period: str
    data: List[TimeSeriesDataPoint]

File: frontend/types/analytics.ts
/**
 * TypeScript interfaces for analytics data.
 */

export interface TimeSeriesDataPoint {
  date: string;
  count: number;
  value?: number;
}

export interface UserStats {
  total_users: number;
  new_users_today: number;
  new_users_this_week: number;
  new_users_this_month: number;
  active_users_today: number;
  active_users_this_week: number;
  active_users_this_month: number;
}

export interface ActivityStats {
  total_activities: number;
  activities_today: number;
  activities_this_week: number;
  activities_this_month: number;
  activities_by_action: Record<string, number>;
  activities_by_resource_type: Record<string, number>;
}

export interface OrganizationStats {
  total_organizations: number;
  total_memberships: number;
  average_members_per_org: number;
  organizations_created_today: number;
  organizations_created_this_week: number;
  organizations_created_this_month: number;
}

export interface ActivityTimeSeries {
  period: string;
  data: TimeSeriesDataPoint[];
}

export interface UserGrowthTimeSeries {
  period: string;
  data: TimeSeriesDataPoint[];
}

export interface DashboardStats {
  user_stats: UserStats;
  activity_stats: ActivityStats;
  organization_stats: OrganizationStats;
  activity_timeline: ActivityTimeSeries;
}

